Fast dataset resume

mariosasko opened the pull request:
This PR makes resuming dataset iteration from a checkpoint fast again.

This performance regression comes from https://github.com/pytorch/torchtitan/pull/838. In that PR, `.skip` is removed for both map-style and iterable-style datasets for correctness reasons. However, `.skip` works as expected for map-style datasets, so the change can be reverted for that case. On the other hand, for iterable-style datasets, calling `.skip` after `split_dataset_by_node` splits the number of elements to skip **across the ranks** (e.g. calling `.skip(10)` after `split_dataset_by_node(<rank>, 2)` effectively skips  5 (`10 // 2 = 5`) elements on each rank), which isn'r what we want/expect, so removing `.skip` was justified there. Still, we can make the whole thing much faster using the [`state_dict` API](https://huggingface.co/docs/datasets/v3.5.0/en/stream#save-a-dataset-checkpoint-and-resume-iteration) for iterable-style datasets, which avoids re-iterating past shards/files when resuming.

tianyu-l commented on tests/unit_tests/test_dataset_checkpointing.py:
```diff
@@ -12,27 +12,32 @@
-        assert torch.equal(input_ids["input"], expected_input_ids["input"])
-        assert torch.equal(labels, expected_labels)
+        for dataset_name in ["c4_test", "c4_test_streaming"]:
+            dataset_name = "c4_test"
```
should remove

tianyu-l commented on torchtitan/datasets/hf_datasets.py:
```diff
@@ -49,6 +49,11 @@ class DatasetConfig:
         loader=lambda path: load_dataset(path, split="train"),
         text_processor=_process_c4_text,
     ),
+    "c4_test_streaming": DatasetConfig(
+        path="tests/assets/c4_test",
+        loader=lambda path: load_dataset(path, split="train", streaming=True),
+        text_processor=_process_c4_text,
+    ),
```
Since this is only used in unit test, can we only register it in the test file?

tianyu-l commented on torchtitan/datasets/hf_datasets.py:
```diff
@@ -97,15 +102,21 @@ def __init__(
         # Variables for checkpointing
         self._sample_idx = 0
         self._all_tokens: list[int] = []
+        self._data_state_dict_loaded = False
```
Thanks for considering backward compatibility. At this moment, I think we should just move with the right and efficient behavior, without worrying too much about it.

tianyu-l commented on torchtitan/datasets/hf_datasets.py:
```diff
@@ -138,8 +149,23 @@ def load_state_dict(self, state_dict):
+        }
+
+        if isinstance(self._data, IterableDataset):
+            _state_dict["data"] = self._data.state_dict()
```
Is this supported in `datasets 2.21.0`, or newer version? If so, we need to update `requirements.txt`.

tianyu-l replied:
Also let's add some comment on the efficiency of this solution with reference to https://huggingface.co/docs/datasets/v3.5.0/en/stream#save-a-dataset-checkpoint-and-resume-iteration

tianyu-l commented on torchtitan/datasets/hf_datasets.py:
```diff
@@ -138,8 +149,23 @@ def load_state_dict(self, state_dict):
         self._all_tokens = state_dict["token_buffer"]
 
+        if isinstance(self._data, IterableDataset):
+            if "data" in state_dict:  # backward compatibility
```
By not worrying about BC breaking, let's assert the existence of `"data"`

tianyu-l commented on torchtitan/datasets/hf_datasets.py:
```diff
@@ -97,15 +102,21 @@ def __init__(
+                return iter([])
+            else:
+                return iter(self._data.skip(self._sample_idx))
+        elif not self._data_state_dict_loaded:  # backward compatibility
+            it = iter(self._data)
+            for _ in range(self._sample_idx):
+                next(it)
+            return it
```
let's remove, for simplicity

tianyu-l commented on torchtitan/datasets/hf_datasets.py:
```diff
@@ -138,8 +149,23 @@ def load_state_dict(self, state_dict):
         self._sample_idx = state_dict["sample_idx"]
         self._all_tokens = state_dict["token_buffer"]
 
+        if isinstance(self._data, IterableDataset):
```
One thing I'm curious is -- why don't map-style datasets provide the same API to make API consistent? It could be as simple as what we do here using `self._data.skip(self._sample_idx)`.

mariosasko replied:
Dunno, I'm no longer involved in the `datasets` project, so I don't know exactly, but I assume the reasoning is that it's not too hard to resume map-style datasets manually, as explained in https://github.com/huggingface/datasets/issues/5454#issuecomment-1956431084

tianyu-l commented on torchtitan/datasets/hf_datasets.py:
```diff
@@ -138,8 +149,23 @@ def load_state_dict(self, state_dict):
     def state_dict(self):
-        return {"token_buffer": self._all_tokens, "sample_idx": self._sample_idx}
+        _state_dict = {
+            "token_buffer": self._all_tokens,
```
Could you do torchtitan a favor and rename `self._all_tokens` to `self._token_buffer`, to make things consistent?

tianyu-l commented on torchtitan/datasets/hf_datasets.py:
```diff
@@ -138,8 +149,23 @@ def load_state_dict(self, state_dict):
-        return {"token_buffer": self._all_tokens, "sample_idx": self._sample_idx}
+        _state_dict = {
+            "token_buffer": self._all_tokens,
+            "sample_idx": self._sample_idx,
```
can we save this in state dict only when it's map-style dataset?

tianyu-l reviewed (commented):
Thank you for making resuming dataset checkpoint efficient! I had some inline comments. Please see if they make sense to you.

Regarding
> On the other hand, for iterable-style datasets, calling `.skip` after `split_dataset_by_node` splits the number of elements to skip **across the ranks** (e.g. calling `.skip(10)` after `split_dataset_by_node(<rank>, 2)` effectively skips  5 (`10 // 2 = 5`) elements on each rank), which isn'r what we want/expect, so removing `.skip` was justified there.

I'm a bit confused by your comment. It sounds that the behavior of `skip` for `IterableDataset` is oblivious of whether it has gone through `split_dataset_by_node` or not, which is not intuitive?

fegin commented:
Can you we have an accuracy verification for this PR? I believe llama3 8B can reproduce the loss issue if the dataset doesn't resume correctly.

mariosasko commented:
Sorry for the delay, I addressed the comments and made the test much more robust. 

> I'm a bit confused by your comment. It sounds that the behavior of skip for IterableDataset is oblivious of whether it has gone through split_dataset_by_node or not, which is not intuitive?

It's hard to explain because the `datasets` logic has bugs [^1]. It should be easier to understand with a [toy example](https://colab.research.google.com/drive/1qhYorItyzDPTcOJIq_-oUa8hciaZNnwo?usp=sharing). What surprised me is that changing the number of the data shards results in a completely different behaviour.

> Can you we have an accuracy verification for this PR? I believe llama3 8B can reproduce the loss issue if the dataset doesn't resume correctly.

I don't have access to A100 / H100 GPUs right now, so it would be great if someone else could do the run. I improved the test, so I'm not sure if this is really needed, though.

[^1]: E.g. notice missing parentheses in [this test](https://github.com/huggingface/datasets/blob/ba451d1d9e99904a13413c087a2b01bf64f0e8d6/tests/test_iterable_dataset.py#L1908-L1912) (in the `datasets` repo) that should catch the problematic behaviour

fegin commented:
Can you also fix the linter error and integration test error? I will try if I can verify with llama3.

mariosasko commented:
Unfortunately, one more bug needs to be fixed, this time directly in `datasets` ...

mariosasko commented:
Okay, I think this is ready for the verification test.

tianyu-l commented on torchtitan/datasets/hf_datasets.py:
```diff
@@ -47,7 +47,7 @@ class DatasetConfig:
         path="tests/assets/c4_test",
         loader=lambda path: load_dataset(path, split="train"),
-        text_processor=_process_c4_text,
+        text_processor=lambda sample: sample["text"],
```
do we need this change?

tianyu-l commented on tests/unit_tests/test_dataset_checkpointing.py:
```diff
@@ -4,35 +4,56 @@
+    def setUp(self):
+        DATASETS["c4_test_streaming"] = DatasetConfig(
+            path="tests/assets/c4_test",
+            loader=lambda path: load_dataset(path, split="train").to_iterable_dataset(
+                num_shards=4
+            ),
```
curious what's the difference between this and setting `streaming=True` in `load_dataset`

mariosasko replied:
With `streaming=True` in `load_dataset` we cannot control the number of underlying shards, but this way we can.

tianyu-l commented on torchtitan/datasets/hf_datasets.py:
```diff
@@ -96,16 +96,16 @@ def __init__(
     def _get_data_iter(self):
-        if isinstance(self._data, Dataset) and self._sample_idx == len(self._data):
-            return iter([])
+        if isinstance(self._data, Dataset):
```
I think it's worth commenting here that map-style `Dataset` and `IterableDataset` have different ways of resuming from checkpoint -- `Dataset` uses the `_sample_idx` in state dict to relocate `self._data` whereas for `IterableDataset` `self._data` has been updated to the correct index during state dict loading.

tianyu-l replied:
also a side note:
It seems we have quite different treatment for map-style and iterable, which causes several if-else conditions. In the future it may be worth a slight refactoring into two subclasses of the same parent sharing common interfaces -- although I still think it's better for `datasets` to provide consistent interface (e.g. removing the need for explicitly storing `_sample_idx`), lol

tianyu-l reviewed (commented):
Looks good to me, had some minor comments.
I'll leave it to @fegin to verify and stamp.

fegin reviewed (approved):
The loss curve looks okay locally.  Please fix the linter. The 8GPU integration test is not related.
